// promptfoo/dist/src/share.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createShareableUrl = void 0;
const fetch_1 = require("./fetch");
const constants_1 = require("./constants");
async function createShareableUrl(results, config) {
    const sharedResults = {
        data: {
            version: 2,
            // TODO(ian): Take date from results, if applicable.
            createdAt: new Date().toISOString(),
            results,
            config,
        },
    };
    const apiBaseUrl = typeof config.sharing === 'object' ? config.sharing.apiBaseUrl : constants_1.REMOTE_API_BASE_URL;
    const response = await (0, fetch_1.fetchWithProxy)(`${apiBaseUrl}/eval`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(sharedResults),
    });
    const { id } = (await response.json());
    const appBaseUrl = typeof config.sharing === 'object' ? config.sharing.appBaseUrl : constants_1.REMOTE_APP_BASE_URL;
    return `${constants_1.REMOTE_APP_BASE_URL}/eval/${id}`;
}
exports.createShareableUrl = createShareableUrl;
