// promptfoo/dist/src/providers/webhook.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookProvider = void 0;
const logger_1 = __importDefault(require("../logger"));
const cache_1 = require("../cache");
const shared_1 = require("./shared");
class WebhookProvider {
    constructor(webhookUrl, options = {}) {
        const { id, config } = options;
        this.webhookUrl = webhookUrl;
        this.id = id ? () => id : this.id;
        this.config = config;
    }
    id() {
        return `webhook:${this.webhookUrl}`;
    }
    toString() {
        return `[Webhook Provider ${this.webhookUrl}]`;
    }
    async callApi(prompt) {
        const params = {
            prompt,
        };
        if (this.config) {
            params.config = this.config;
        }
        logger_1.default.debug(`Calling Webhook: ${this.webhookUrl} with params: ${JSON.stringify(params)}`);
        let response;
        try {
            response = await (0, cache_1.fetchWithCache)(this.webhookUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(params),
            }, shared_1.REQUEST_TIMEOUT_MS, 'json');
        }
        catch (err) {
            return {
                error: `Webhook call error: ${String(err)}`,
            };
        }
        logger_1.default.debug(`\tWebhook response: ${JSON.stringify(response.data)}`);
        if (response.data && typeof response.data.output === 'string') {
            return {
                output: response.data.output,
            };
        }
        else {
            return {
                error: `Webhook response error: Unexpected response format: ${JSON.stringify(response.data)}`,
            };
        }
    }
}
exports.WebhookProvider = WebhookProvider;
